File: apps/api/src/mentions/mention-token.ts
/**
 * Разбор упоминаний в тексте комментария и описания задачи (US-74).
 *
 * **Упоминание хранится ссылкой на пользователя, а не именем.** В тексте оно
 * записано токеном `@[Имя](user:<uuid>)`:
 *  - идентификатор — то, по чему создаётся связь и уведомление. Имя в токене
 *    участвует только в запасном отображении;
 *  - актуальное имя клиент берёт из списка `mentions` в ответе API, поэтому смена
 *    имени в Яндекс ID видна во всех старых комментариях (US-74);
 *  - если человек больше не участник проекта, связи в ответе не будет, и клиент
 *    покажет имя из токена **обычным текстом** — как и требует US-74.
 *
 * Всё, что не совпало с токеном (в том числе набранное вручную `@ivan`), остаётся
 * обычным текстом: ни подсветки, ни связи, ни уведомления (D-41).
 */

/** Токен упоминания. Имя — до 120 символов и без `]`, чтобы разбор был однозначным. */
const MENTION_TOKEN =
  /@\[([^\]\r\n]{1,120})\]\(user:([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})\)/gu;

export interface ParsedMention {
  userId: string;
  /** Имя на момент написания — запасной вариант отображения, не источник правды. */
  displayName: string;
}

/**
 * Все упоминания текста, каждый пользователь — один раз (US-74: повторное упоминание
 * того же человека в одном комментарии создаёт одно уведомление).
 */
export function parseMentions(body: string): ParsedMention[] {
  const seen = new Map<string, ParsedMention>();

  for (const match of body.matchAll(MENTION_TOKEN)) {
    const userId = match[2]!.toLowerCase();
    if (!seen.has(userId)) {
      seen.set(userId, { userId, displayName: match[1]!.trim() });
    }
  }

  return [...seen.values()];
}

/** Текст без токенов: `@[Анна](user:…)` → `@Анна`. Нужен превью уведомления и поиску. */
export function renderMentionsAsText(body: string): string {
  return body.replace(MENTION_TOKEN, (_match, name: string) => `@${name.trim()}`);
}

/** Собрать токен — нужен тестам и клиентским примерам в документации. */
export function mentionToken(userId: string, displayName: string): string {
  return `@[${displayName}](user:${userId})`;
}

File: apps/web/lib/core/domain/mention_token.dart
import 'package:flutter/foundation.dart';

/// Упомянутый человек: то, чем упоминание рисуется на экране.
///
/// Отдельный тип, а не `IssueUserDto`, потому что упоминания рендерит виджет
/// дизайн-системы, а он про контракт API знать не должен.
@immutable
class MentionRef {
  /// @nodoc
  const MentionRef({
    required this.id,
    required this.displayName,
    this.avatarUrl,
  });

  /// Идентификатор пользователя — он же то, что стоит в токене.
  final String id;

  /// **Актуальное** имя: человек мог сменить его в Яндекс ID уже после того,
  /// как комментарий написали (US-74).
  final String displayName;

  /// @nodoc
  final String? avatarUrl;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is MentionRef &&
          other.id == id &&
          other.displayName == displayName &&
          other.avatarUrl == avatarUrl;

  @override
  int get hashCode => Object.hash(id, displayName, avatarUrl);
}

/// Токен упоминания в тексте: `@[Имя](user:<uuid>)`.
///
/// Источник правды — **идентификатор**, имя внутри токена запасное
/// (US-74, D-41). Отсюда два правила рендера:
///
/// 1. Имя берётся из поля `mentions` ответа, а не из текста: тогда смена
///    имени видна во всех старых комментариях сразу.
/// 2. Токен, которому в `mentions` ничего не соответствует, показывается
///    **обычным текстом** — это упоминание постороннего, сервер его молча
///    проигнорировал, и притворяться, что связь есть, нельзя.
sealed class MentionToken {
  /// Разметка токена.
  ///
  /// Имя не содержит `]` — так его вставляет подсказка, и так его разбирает
  /// сервер. Идентификатор — UUID, без строгой проверки версии: клиент
  /// проверяет форму, а существование — сервер.
  static final pattern = RegExp(
    r'@\[([^\]\n]*)\]\(user:([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}'
    r'-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})\)',
  );

  /// Собирает токен для вставки в текст.
  ///
  /// Имя чистится от символов, которые сломали бы разбор: остальное — как
  /// его прислала подсказка.
  static String format({required String id, required String displayName}) {
    final safeName = displayName.replaceAll(RegExp(r'[\[\]\n]'), ' ').trim();

    return '@[$safeName](user:$id)';
  }

  /// Заменяет токены на «@Имя» — для мест, где разметки нет вовсе:
  /// доступное имя, черновик в свёрнутом поле, заголовок уведомления.
  ///
  /// [names] — актуальные имена по идентификаторам. Незнакомый токен
  /// остаётся как есть: он и должен читаться как обычный текст.
  static String toPlainText(
    String body, {
    Map<String, String> names = const {},
  }) => body.replaceAllMapped(pattern, (match) {
    final id = match.group(2)!;
    final actual = names[id];

    return actual == null ? match.group(0)! : '@$actual';
  });
}
